File: Linked List/61. Rotate List/61. Rotate List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// 本题的核心思想较为简单：假设链表的结点总数为size，实际上就是找到链表的第k%size个结点，然后将链表从此处断开，前半段链表接到后半段链表之后，更新头结点指针

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        // 0. 特别判断链表是否为空
        if(!head)
        return head;

        // 1. 首先计算链表的总结点数量
        int size=0;
        ListNode *pos=head;    //链表游标指针
        while(pos)        
        {
            pos=pos->next;
            size+=1;
        }

        // 2. 计算实际应该旋转的结点数量，实际应该旋转的结点数量为k=k%size
        k=k%size;
        
        if(!k)    // 特别处理，若实际应该旋转的结点数量为0，则无需进行旋转
        return head;

        // 3. 需要旋转的结点数量为k%size，则我们使用双指针法找到倒数第k%size+1个结点
        ListNode *fast=head;   //快指针
        ListNode *slow=head;   //慢指针

        for(int i=0;i<k+1;i++)   
        {
            fast=fast->next;   //快指针先移动k+1次
        }

        while(fast)            //然后同时移动快慢指针，直到快指针移动到链表末尾NULL处，这时慢指针就指向倒数第k+1个结点
        {
            fast=fast->next;
            slow=slow->next;
        }

        // 4. 将链表从指针slow处分为前后两段，然后将前半段连接到后半段后
        ListNode *back=slow->next;     //链表后半段
        ListNode *front=head;          //链表前半段
        ListNode *backpos=back;        //链表后半段指针

        slow->next=NULL;              //链表前半段封口（易错点）
        
        //查找后半段的最后一个结点
        while(backpos->next)
        {
            backpos=backpos->next;
        }

        //将前半段连接到后半段后
        backpos->next=front;

        return back;
    }
};

File: Linked List/61. Rotate List/61. Rotate List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        # 0. 处理链表为空的情况(但凡链表题均需要优先注意该情况)
        if not head:
            return head

        # 1. 首先计算链表的结点总数
        size=0
        pos=head
        while pos:
            pos=pos.next
            size+=1
        
        # 2. 计算需要旋转的结点数目
        k=k%size

        if not k:     # 特殊情况: k==0时相当于没有结点需要进行旋转，此时直接返回原链表
            return head

        # 3. 使用快慢双指针法查找链表的倒数第k%size+1个结点
        fast=head
        slow=head

        for i in range(k+1):
            fast=fast.next
        
        while fast:
            fast=fast.next
            slow=slow.next
        
        # 4. 此时slow即为倒数的第k%size+1个结点，从该结点开始，将链表分为前后两个部分
        front=head        # 前半段链表
        back=slow.next    # 后半段链表
        backpos=back      # 后半段链表指针

        slow.next=None    # 前半段链表封口

        while backpos.next:
            backpos=backpos.next

        # 5. 将前半段链表连接到后半段链表后

        backpos.next=front

        return back
